File: sql-insight-engine/sql-insight-engine-innodb/src/main/java/org/gongxuanzhang/sql/insight/core/engine/storage/innodb/page/Infimum.java
/*
 * Copyright 2023 sql-insight  and the original author or authors <gongxuanzhangmelt@gmail.com>.
 *
 * Licensed under the GNU Affero General Public License v3.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://github.com/implement-study/sql-insight/blob/main/LICENSE
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.gongxuanzhang.sql.insight.core.engine.storage.innodb.page;


import org.gongxuanzhang.easybyte.core.DynamicByteBuffer;
import org.gongxuanzhang.sql.insight.core.engine.storage.innodb.page.compact.RecordHeader;
import org.gongxuanzhang.sql.insight.core.object.Table;
import org.gongxuanzhang.sql.insight.core.object.value.Value;

import java.util.List;


/**
 * min record in group
 *
 * @author gxz gongxuanzhangmelt@gmail.com
 **/
public class Infimum implements InnodbUserRecord {

    private static final String INFIMUM_BODY = "infimum";

    /**
     * 5 bytes.
     **/
    RecordHeader recordHeader;

    /**
     * fixed 8 bytes. "infimum" is 7 bytes . fill 0 zero occupy the space
     **/
    final byte[] body;

    public Infimum() {
        body = DynamicByteBuffer.wrap(INFIMUM_BODY.getBytes()).append((byte) 0).toBytes();
    }

    @Override
    public byte[] rowBytes() {
        return DynamicByteBuffer.wrap(recordHeader.toBytes()).append(this.body).toBytes();
    }

    @Override
    public int nextRecordOffset() {
        return recordHeader.getNextRecordOffset();
    }

    @Override
    public boolean deleteSign() {
        return false;
    }

    @Override
    public int offset() {
        return ConstantSize.INFIMUM.offset();
    }


    @Override
    public String toString() {
        return this.recordHeader.toString() + "[body:" + new String(this.body) + "]";
    }

    @Override
    public List<Value> getValues() {
        return infimumUnsupport();
    }

    @Override
    public long getRowId() {
        return infimumUnsupport();
    }

    @Override
    public Value getValueByColumnName(String colName) {
        return infimumUnsupport();
    }

    @Override
    public Table belongTo() {
        return infimumUnsupport();
    }

    private <T> T infimumUnsupport() {
        throw new UnsupportedOperationException("this is infimum!");
    }

    @Override
    public RecordHeader getRecordHeader() {
        return this.recordHeader;
    }

    @Override
    public int length() {
        return ConstantSize.INFIMUM.size();
    }
}

File: sql-insight-engine/sql-insight-engine-innodb/src/main/kotlin/tech/insight/engine/innodb/page/Infimum.kt
package tech.insight.engine.innodb.page

import org.gongxuanzhang.easybyte.core.DynamicByteBuffer
import tech.insight.core.bean.Row
import tech.insight.core.bean.Table
import tech.insight.core.bean.value.Value
import tech.insight.engine.innodb.factory.RecordHeaderFactory
import tech.insight.engine.innodb.page.compact.RecordHeader


/**
 * min record in group
 *
 * @author gxz gongxuanzhangmelt@gmail.com
 */
class Infimum : InnodbUserRecord {

    companion object {
        private const val INFIMUM_BODY = "infimum"
    }

    /**
     * 5 bytes.
     */
    override var recordHeader: RecordHeader

    /**
     * fixed 8 bytes. "infimum" is 7 bytes . fill 0 zero occupy the space
     */
    val body: ByteArray

    init {
        recordHeader = RecordHeaderFactory.infimumHeader()
        body = DynamicByteBuffer.wrap(INFIMUM_BODY.toByteArray()).append(0.toByte()).toBytes()
    }

    override fun rowBytes(): ByteArray {
        return DynamicByteBuffer.wrap(recordHeader.toBytes()).append(body).toBytes()
    }

    override fun beforeSplitOffset(): Int {
        return recordHeader.length()
    }

    override fun nextRecordOffset(): Int {
        return recordHeader.nextRecordOffset
    }

    override fun deleteSign(): Boolean {
        return false
    }

    override fun offset(): Int {
        return ConstantSize.INFIMUM.offset()
    }

    override fun setOffset(offset: Int) {
        throw UnsupportedOperationException("infimum can't set offset ")
    }

    override fun toString(): String {
        return recordHeader.toString() + "[body:" + String(body) + "]"
    }

    override val values: List<Value<*>>
        get() = infimumUnsupported()
    override val rowId: Long
        get() = infimumUnsupported()

    override fun getValueByColumnName(colName: String): Value<*> {
        infimumUnsupported()
    }

    override fun belongTo(): Table {
        infimumUnsupported()
    }

    override operator fun compareTo(that: Row): Int {
        return if (that is InnodbUserRecord) {
            -1
        } else infimumUnsupported()
    }

    private fun infimumUnsupported(): Nothing {
        throw UnsupportedOperationException("this is infimum!")
    }


    override fun length(): Int {
        return ConstantSize.INFIMUM.size()
    }

    fun setRecordHeader(recordHeader: RecordHeader): Infimum {
        this.recordHeader = recordHeader
        return this
    }


}
